TACC/Core-Portal
# Task/WP-80 Implement dynamic execution system selection

## Overview
There are exec system agnostic apps, and the Portal Job submission form needs to allow the users to pick an execution system for such apps. This PR implements the feature to provide that.

### Implementation decisions:
* UI Visibility of execution systems:
  * Visibility of execution system is dependent on allocation. 
  * By default if a portal has allocation it is selected as default.
  * Next, the list of execution systems is dependent on the what's available in the allocation.
  
  The alternative is to show all execution systems available to user, which is not going to work well because user will blocked by allocation.

* Node and Time limits when exec system is flipped:
  When exec system selection is flipped in the form, should the node and time limits match what's in the app defaults or should it keep the user modifications to those defaults.
   * Bring the app defaults:
      Steps: 
      1. user selects frontera, the app default of node count:1 is shown
      2. user updates the queue to non-default (development)
      3. node count updates to 40
      4. user select ls6
      After step 4, node count goes back to 1 and default queue for ls6.

   * Keep the current node count on changing exec system (the PR implemented this)
     Steps: 
      1. user selects frontera, the app default of node count:1 is shown
      2. user updates the queue to non-default (development)
      3. node count updates to 40
      4. user select ls6, queue updates to ls6 default
      After step 4, node count adjust to ls6 default queue and will not go back to 1.

* Cannot use tapis job attribute dynamicExecSystem as an indicator of whether app supports dynamic exec system.
   * [Tapis doc note](https://tapis.readthedocs.io/en/latest/technical/apps.html): 
 Currently dynamic selection of an execution system is not supported. For this reason the job related attribute dynamicExecSystem should be set to false (the default) and execSystemConstraints should not be set.
  * Alternative is to use notes section in the app and add a boolean attribute `dynamicExecSystem` 


## Related

* [WP-80](https://tacc-main.atlassian.net/browse/WP-80)

## Changes

### Backend:

* Dependent picklist
  * Execution System is dependent on allocation selection.
  * Queue and node limits are dependent on Execution System (existing feature, just integrated into execution system selection).
  * Make necessary app form state handling to keep tracking of picklist changes.

* Add a notes option for each app, to allow a selection of dynamic execution systems. Existance of this attribute indicates that dynamic execution system feature is available. There are two ways to provide dynamic execution systems.
  * A list of systems. This is the list of restricted systems that this app can execute.
  * "All". List of all systems that the user has access to.
  * Handle both "All" vs "specific list". Send the execution systems for the app user when the notes indicates dynamic execution system.

  Example:

* Specific list of execution systems. Only LoneStar6, Frontera are shown if the user has access to them.

```
notes: 
...
dynamicExecSystems: ["ls6", "frontera"]
...
```
* All available execution systems
```
notes: 
...
dynamicExecSystems: ["All"]
...
```
* Without dynamic execution Systems. This will use execution system in job attribute.
```
notes: 
label: "namd",
...
```

* Adjust exec system attribute - change from exec_sys to array `execSystems`.

### Client:
* Move allocation drop down to the top of the configuration.
* Make execution system drop down dependent on the allocation.
* Setup a drop down for exec system.
* Handle changes when the exec system selection is updated. 
   * Update current component to a generic name.
   * Update queue and node values. 
   * Create a state for the form and update it.
* Allocation changes:
   * Track allocation per exec system.
   * When exec system, trigger the allocation check.
* Label for execution system
   * System definition now has label. This label is used.
   * Otherwise, system name is used for execution system.

## Testing




1) App with no dynamicExecSystem in notes section:
   - No regression
      - job submission works
      - changing queue will adjust other aspects in the form (node and minutes)
      - default queue is properly selected and form values are correct.
      - Job history page shows right exec system information.
   - Existing automated tests work.
2) App with dynamic execution system
   - Pre-req:
      - Hello world or similar app with notes added for  `dynamicExecSystems: ["ls6", "frontera"]`
      - Create exec system in tapis v3 (you can use your own user to do it, so the list has some use cases for testing). This will make testing more interesting.
  
   - Flip between frontera and ls6. Check the following:
      - job submission works
      - changing queue will adjust other aspects in the form (node and minutes)
      - default queue is properly selected and form values are correct.
      - Job history page shows right exec system information.
   - Use an allocation with no available execution systems and see what happens. (I only saw 3DEM based allocation with no hosts).
    

3) Test automation was added/updated for the following scenarios:

    * No dynamic exec system enabled - update existing tests for validation
    * App with dynamic stuff
   
     - all exec systems with allocation
     - some exec systems with no allocation
     - only one exec system with allocation
     - only one exec system with allocation
     - no exec system with allocation
     
     - 2 exec systems with missing default queue
     - 2 exec systems with different queue names
     - 2 exec systems - queue selection and changing exec system
     - 2 exec systems - modify queue selection and check the corresponding dependencies
     - 2 exec systems - modify queue limits and check if they are set appropriately in the values
     - 2 exec systems - modify queue on new exec system
     
     Verification:
     - if exec system drop down exists, the values are right
     - the queue values are right
     - the node and rest of the selection value are right

## UI

* Order of drop down is different. Allocation is first and then Systems.
![Screenshot 2024-05-03 at 10 13 08 AM](https://github.com/TACC/Core-Portal/assets/2568355/3a957266-db77-431f-b396-8b38e178ba23)

* Added drop down to show exec system with labels from system.
![Screenshot 2024-05-03 at 10 13 17 AM](https://github.com/TACC/Core-Portal/assets/2568355/6625ea73-971f-4b6f-84bf-ca9b3c61ec55)

* Error message when allocation has no execution system.
<img width="801" alt="Screenshot 2024-06-12 at 9 37 37 AM" src="https://github.com/TACC/Core-Portal/assets/2568355/082c6429-15b8-4c29-b9ef-33225fb2db2b">

* Error message when user further continues other selection without adjusting allocation
<img width="777" alt="Screenshot 2024-06-12 at 9 37 43 AM" src="https://github.com/TACC/Core-Portal/assets/2568355/7e4c2115-4308-4bf3-b157-6f5059b0b642">


* Job History display of System Name

![Screenshot 2024-05-03 at 10 21 20 AM](https://github.com/TACC/Core-Portal/assets/2568355/f599eb71-b52a-4f08-ba47-2f78da36b6d3)



## Notes
